bitfield/bitfield.go

```diff
@@ -0,0 +1,101 @@
+package hypercore
+
+import (
+	"github.com/kiambogo/go-hypercore/mempager"
+)
+
+type Bitfield struct {
+	pager      *mempager.Pager
+	byteLength uint64
+}
+
+func NewBitfield(pageSize int) *Bitfield {
+	pgr := mempager.NewPager(pageSize)
+	return &Bitfield{pager: &pgr}
+}
+
+// PageSize returns the size of the pages used by the internal pager
+func (b Bitfield) PageSize() int {
+	return b.pager.PageSize()
+}
+
+// ByteLength returns the number of bytes in the bitfield
+func (b Bitfield) ByteLength() uint64 {
+	return b.byteLength
+}
+
+// Len returns the number of bits set in the bitfield
+func (b Bitfield) Len() uint64 {
+	return b.byteLength * 8
+}
+
+// IsEmpty returns true if no bits are stored in the bitfield
+func (b Bitfield) IsEmpty() bool {
+	return b.pager.IsEmpty()
+}
+
+// SetBit sets the bit at a particular index within the bitfield
+// Returns true if a change was inacted
+func (b *Bitfield) SetBit(index int, value bool) bool {
+	byteIndex := uint64(index / 8) // 8 bits in a byte
+	byteAtOffset := b.GetByte(byteIndex)
+
+	bitIndex := byte(1 << (index % 8))
+
+	if bitAlreadySet := (byteAtOffset&bitIndex == 1) == value; bitAlreadySet {
```

Comment:
On this bit, I'm wondering if this complexity to understand is worth the early return here versus potentially:
a) An easier comparison after calculating the `updatedByte` below, like `if updatedByte == byteAtOffset` or,
b) leaving this all up to `SetByte` which does the comparison anyway.

Might be worth some benchmarking fun

Reply:
yep, I had a similar thought. I opted for the early return to take advantage of the slight performance gain by _not_ executing `SetByte()` in the case that the bit is already set. Based on your benchmarking (thanks!) it seems like it might not be too much of a perf gain, although that can vary on the usage (where your benchmark uses rand(100000)). 

In this case, I think a) makes the code read a bit clearer, so I'll go for that. good discussion! 🌮